File: frontend/src/app/core/services/car.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CarRequest, CarResponse } from '../models/car.model';

/**
 * service for interacting with the car listings api endpoints.
 */
@Injectable({
  providedIn: 'root'
})
export class CarService {
  private apiUrl = '/api/cars';

  constructor(private http: HttpClient) {}

  getAllActiveCars(): Observable<CarResponse[]> {
    return this.http.get<CarResponse[]>(this.apiUrl);
  }

  getCarById(id: number): Observable<CarResponse> {
    return this.http.get<CarResponse>(`${this.apiUrl}/${id}`);
  }

  getMyCars(): Observable<CarResponse[]> {
    return this.http.get<CarResponse[]>(`${this.apiUrl}/my-cars`);
  }

  createCar(car: CarRequest): Observable<CarResponse> {
    return this.http.post<CarResponse>(this.apiUrl, car);
  }

  updateCar(id: number, car: CarRequest): Observable<CarResponse> {
    return this.http.put<CarResponse>(`${this.apiUrl}/${id}`, car);
  }

  deleteCar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: backend/src/main/java/com/example/demo/service/CarService.java
package com.example.demo.service;

import com.example.demo.dto.request.CarRequest;
import com.example.demo.dto.response.CarResponse;

import java.util.List;

/**
 * service interface for handling car-related business operations.
 */
public interface CarService {

    CarResponse createCar(CarRequest request, String ownerUsername);

    CarResponse updateCar(Long id, CarRequest request, String currentUsername);

    void deleteCar(Long id, String currentUsername);

    List<CarResponse> getAllActiveCars();

    List<CarResponse> getCarsByOwner(String username);

    CarResponse getCarById(Long id);
}
